torch/testing/_internal/common_utils.py

```diff
@@ -807,10 +861,818 @@ def wrapper(*args, **kwargs):
+                torch._C._set_conj(r, t.is_conj())
+                torch._C._set_neg(r, t.is_neg())
+            return r
+
+        def to_meta(t):
+            nonlocal hit, miss
+            # TODO: zero tensors?  We appear to have eliminated them by
+            # excluding complex for now
+            if type(t) is torch.Tensor or type(t) is torch.nn.Parameter:
+                if any([
+                    t.is_sparse_csr, t.is_sparse, t.is_mkldnn, t.is_quantized,
+                    t.is_nested, torch._is_functional_tensor(t),
+                    # these are supported in meta conversion but the fallbacks
+                    # don't work
+                    t.is_neg(), t.is_conj(),
+                    # conjugate fallback does not support meta tensors
+                    t.dtype in (torch.complex128, torch.complex64),
+                ]):
+                    # TODO: sparse should support meta
+                    # NB technically to('meta') does work but our logging
+                    # instrumentation will see the meta conversions and the
+                    # tests all break so we just exclude this.  In any case
+                    # the to conversion isn't really right anyhow.
+                    miss += 1
+                    return t
+                elif any([
+                    t.device.type in ("lazy", "meta"), t.is_complex(),
+                    torch._C._is_batched(t),
+                ]):
+                    # TODO: this stuff should support storage
+                    # (well, maybe not batched)
+                    hit += 1
+                    return t.to("meta")
+                else:
+                    hit += 1
+                    r = meta_tensor(t)
+                    if type(t) is torch.nn.Parameter:
+                        r = torch.nn.Parameter(r, requires_grad=r.requires_grad)
+                    return r
+            elif torch.overrides.is_tensor_like(t):
+                # Blindly converting tensor subclasses to meta can cause
+                # unpredictable problems; e.g., FX tests will trace meta
+                # tensors into their trace / some subclasses don't correctly
+                # support meta.  Trying to YOLO this is more trouble than it's
+                # worth.
+                miss += 1
+                return t
+            else:
+                # non-Tensor types don't count as hit or miss
+                return t
+
+        do_meta = (
+            func not in skipped and
+            func not in meta_exclude_set and not torch.jit.is_tracing() and
+            not isinstance(func, torch.ScriptMethod)
+        )
+
+        if do_meta:
+            try:
+                meta_args = tree_map(to_meta, args)
+                meta_kwargs = tree_map(to_meta, kwargs)
+            except Exception as e:
+                raise RuntimeError(
+                    f"failed to convert args to meta; "
+                    f"originally (*{args}, **{kwargs})") from e
```

Comment:
High-level question: is crossref meant to test only meta, or will it also test other things (like torch-> prim decompositions, aten->??? decompositions) in the future?

If it is the latter: right now there is a single skipIfCrossRef decorator, we would want different decorators for different type of crossref tests

Reply:
Yes, this is probably right. Some of the tests I'm skipping probably should be permanently skipped for all cross refs because they're very painful (cpp extensions sticks out). Actually, some of these skips may be possible to remove again since I tightened cross ref conditions